desc/integrals/bounce_utils.py

```diff
@@ -0,0 +1,809 @@
+        Shape must broadcast with (P, S).
+        λ values to evaluate the bounce integral at each field line. λ(ρ,α) is
+        specified by ``pitch[...,(ρ,α)]`` where in the latter the labels (ρ,α) are
+        interpreted as the index into the last axis that corresponds to that field
+        line. If two-dimensional, the first axis is the batch axis.
+
+    """
+    errorif(knots.ndim != 1, msg=f"knots should be 1d; got shape {knots.shape}.")
+    errorif(
+        g.shape[-1] != (knots.size - 1),
+        msg=(
+            "Last axis does not enumerate polynomials of spline. "
+            f"Spline shape {g.shape}. Knots shape {knots.shape}."
+        ),
+    )
+    errorif(
+        g.ndim > 3
+        or dg_dz.ndim > 3
+        or (g.shape[0] - 1) != dg_dz.shape[0]
+        or g.shape[1:] != dg_dz.shape[1:],
+        msg=f"Invalid shape {g.shape} for spline and derivative {dg_dz.shape}.",
+    )
+    # Add axis which enumerates field lines if necessary.
+    g, dg_dz = atleast_3d_mid(g, dg_dz)
+    if pitch is not None:
+        pitch = jnp.atleast_2d(pitch)
+        errorif(
+            pitch.ndim != 2
+            or not (pitch.shape[-1] == 1 or pitch.shape[-1] == g.shape[1]),
+            msg=f"Invalid shape {pitch.shape} for pitch angles.",
+        )
+    return g, dg_dz, pitch
+
+
+def bounce_points(
+    pitch, knots, B, dB_dz, num_well=None, check=False, plot=True, **kwargs
+):
+    """Compute the bounce points given spline of |B| and pitch λ.
+
+    Parameters
+    ----------
+    pitch : jnp.ndarray
+        Shape must broadcast with (P, S).
+        λ values to evaluate the bounce integral at each field line. λ(ρ,α) is
+        specified by ``pitch[...,(ρ,α)]`` where in the latter the labels (ρ,α) are
+        interpreted as the index into the last axis that corresponds to that field
+        line. If two-dimensional, the first axis is the batch axis.
+    knots : jnp.ndarray
+        Shape (knots.size, ).
+        ζ coordinates of spline knots. Must be strictly increasing.
+    B : jnp.ndarray
+        Shape (B.shape[0], S, knots.size - 1).
+        Polynomial coefficients of the spline of |B| in local power basis.
+        First axis enumerates the coefficients of power series. Second axis
+        enumerates the splines. Last axis enumerates the polynomials that
+        compose a particular spline.
+    dB_dz : jnp.ndarray
+        Shape (B.shape[0] - 1, *B.shape[1:]).
+        Polynomial coefficients of the spline of (∂|B|/∂ζ)|(ρ,α) in local power basis.
+        First axis enumerates the coefficients of power series. Second axis
+        enumerates the splines. Last axis enumerates the polynomials that
+        compose a particular spline.
+    num_well : int or None
+        Specify to return the first ``num_well`` pairs of bounce points for each
+        pitch along each field line. This is useful if ``num_well`` tightly
+        bounds the actual number. As a reference, there are typically at most 5
```

Comment:
In general, the maximum number of wells is  M + ι N.
This is the maximum frequency along the field line of the Fourier representation of |B|
in straight field line coordinates.

g : ϑ, ϕ ↦ ∑ₘₙ aₘₙ exp(j [mϑ + nϕ])
g : α, ϕ ↦ ∑ₘₙ aₘₙ exp(j [mα + (m ι + n)ϕ])

It's hard to say what  M + ι N is in terms of DESC's spectral resolution, because I don't know how the stream function condenses the fourier spectrum.

From plotting our examples, such as Heliotron, I have found this number to be typically 20. (I have updated from 5 to 20 and shared how users can check for themselves).

Reply:
Thinking about this more. As a rule of thumb, the spectral width of B is ~2x the spectral width of R, Z, and in SFL coordinates like PEST you probably need M,N 2x larger than in DESC coordinates (or even more in boozer, maybe 3-5x). So assuming iota ~ 1 this would put a max of ~4(M+N) though in practice probably a lot less.